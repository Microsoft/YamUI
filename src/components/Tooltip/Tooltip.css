/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";
@import "src/css/variables/fonts.css";

.y-tooltip--callout {
  box-shadow: none !important;
  padding: .7rem 1.2rem .8rem;
  z-index: 1000;
  min-height: 32px;
}

.y-tooltip .ms-Tooltip-content {
  @extend text__small;
  color: $textColor__white;
}
.y-tooltip--callout > .ms-Callout-main,
.y-tooltip--callout > .ms-Callout-beak,
.y-tooltip--callout > .ms-Callout-beakCurtain {
  background-color: $textColor;
}

.y-tooltip--callout > .ms-Callout-beakCurtain {
  border-radius: 2px;
}
