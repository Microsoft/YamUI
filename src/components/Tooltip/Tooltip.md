### Notes for use
You should always wrap Tooltip around a tabbable element, ensuring keyboard navigation can focus on a nested element to trigger the tooltip.

### Examples

Basic usage:

```js { "props": { "data-description": "basic" } }
<Tooltip text="This is the tooltip content.">
  <Clickable>Hover over me</Clickable>
</Tooltip>
```

Dark Background and block-level child:

```js { "props": { "data-description": "dark background" } }
<div style={{ backgroundColor: '#666', padding: '50px 15px 15px 50px' }}>
  <Tooltip text="This is the tooltip content.">
    <Button text='Button' />
  </Tooltip>
</div>
```

Very long text:

```js { "props": { "data-description": "long text" } }
const text = "The quick brown fox jumped over the lazy dog.";

<Tooltip text={`${text} ${text} ${text} ${text} ${text}`}>
  <Clickable>Hover over me</Clickable>
</Tooltip>
```

With directionalHint:

```js { "props": { "data-description": "with directional hint" } }
const text = "The quick brown fox jumped over the lazy dog.";

<Tooltip text={text} directionalHint={3}>
  <Clickable>Hover over me</Clickable>
</Tooltip>
```
