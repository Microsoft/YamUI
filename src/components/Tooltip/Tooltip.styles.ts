/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { palette } from '../../util/colors';
import { TooltipProps } from './Tooltip';
import { borderRadiusSoft } from '../../util/styles';
import { ITooltipStyles } from 'office-ui-fabric-react/lib/components/Tooltip/Tooltip.types';
import { ICalloutContentStyles } from 'office-ui-fabric-react/lib/components/Callout/Callout.types';
import { lineHeights, fontSizes } from '../Text/Text.styles';

export const getTooltipStyles = (_: TooltipProps): ITooltipStyles => {
  return {
    root: {},
    content: {
      display: 'inline-block',
    },
    subText: {
      lineHeight: lineHeights.small,
      fontSize: fontSizes.small,
      color: palette.white,
    },
  };
};

export const getCalloutStyles = (_: TooltipProps): ICalloutContentStyles => {
  return {
    container: {
      backgroundColor: 'inherit',
    },
    root: {
      backgroundColor: 'inherit',
      boxShadow: 'none',
      padding: '0.7rem 1.2rem 0.8rem',
      minHeight: '3.2rem',
      zIndex: 1000,
      border: 'none',
    },
    beak: {
      backgroundColor: palette.neutralPrimary,
    },
    beakCurtain: {
      backgroundColor: palette.neutralPrimary,
      ...borderRadiusSoft,
    },
    calloutMain: {
      backgroundColor: palette.neutralPrimary,
    },
  };
};
