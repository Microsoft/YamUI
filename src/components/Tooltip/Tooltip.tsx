/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { join } from '../../util/classNames';
import { DirectionalHint, TooltipDelay, TooltipHost } from 'office-ui-fabric-react/lib/Tooltip';
import './Tooltip.css';


export { DirectionalHint };

export interface TooltipProps extends NestableBaseComponentProps {
  /**
   * Text content to display in the tooltip.
   */
  text: string;

  /**
   * Preferred direction to show the tooltip in.
   */
  directionalHint?: DirectionalHint;
}

/**
 * Tooltip wraps other content and provides additional context on hover or keyboard selection.
 */
export default class Tooltip extends React.Component<TooltipProps, {}> {
  render() {
    const { className, children, directionalHint, text } = this.props;

    return (
      <TooltipHost
        content={text}
        directionalHint={directionalHint}
        delay={TooltipDelay.medium}
        calloutProps={{ doNotLayer: true, preventDismissOnScroll: true }}
        hostClassName={join(['y-tooltip', className])}
        className="y-tooltip--callout"
      >
        {children}
      </TooltipHost>
    );
  }
}
