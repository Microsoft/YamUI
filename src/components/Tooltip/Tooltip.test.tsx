/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Tooltip, { DirectionalHint, TooltipProps } from './index';

describe('<Tooltip />', () => {
  let component: ShallowWrapper<TooltipProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<Tooltip text="text content">text</Tooltip>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<Tooltip text="text content" className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with directionalHint', () => {
    beforeEach(() => {
      component = shallow(<Tooltip text="text content" directionalHint={DirectionalHint.bottomLeftEdge} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
