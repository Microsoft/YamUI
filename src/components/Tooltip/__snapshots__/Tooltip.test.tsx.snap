// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> with additional className matches its snapshot 1`] = `
<TooltipHost
  calloutProps={
    Object {
      "beakWidth": 8,
      "styles": Object {
        "beak": Object {
          "backgroundColor": "#343A41",
        },
        "beakCurtain": Object {
          "backgroundColor": "#343A41",
          "borderRadius": "2px",
        },
        "calloutMain": Object {
          "backgroundColor": "#343A41",
        },
        "container": Object {
          "backgroundColor": "inherit",
        },
        "root": Object {
          "backgroundColor": "inherit",
          "border": "none",
          "boxShadow": "none",
          "minHeight": "3.2rem",
          "padding": "0.7rem 1.2rem 0.8rem",
          "zIndex": 1000,
        },
      },
    }
  }
  content="text content"
  delay={1}
  hostClassName="y-tooltip TEST_CLASSNAME"
  tooltipProps={
    Object {
      "styles": Object {
        "content": Object {
          "display": "inline-block",
        },
        "root": Object {
          "lineHeight": "1.6rem",
        },
        "subText": Object {
          "color": "#ffffff",
          "fontSize": "1.2rem",
        },
      },
    }
  }
/>
`;

exports[`<Tooltip /> with default options matches its snapshot 1`] = `
<TooltipHost
  calloutProps={
    Object {
      "beakWidth": 8,
      "styles": Object {
        "beak": Object {
          "backgroundColor": "#343A41",
        },
        "beakCurtain": Object {
          "backgroundColor": "#343A41",
          "borderRadius": "2px",
        },
        "calloutMain": Object {
          "backgroundColor": "#343A41",
        },
        "container": Object {
          "backgroundColor": "inherit",
        },
        "root": Object {
          "backgroundColor": "inherit",
          "border": "none",
          "boxShadow": "none",
          "minHeight": "3.2rem",
          "padding": "0.7rem 1.2rem 0.8rem",
          "zIndex": 1000,
        },
      },
    }
  }
  content="text content"
  delay={1}
  hostClassName="y-tooltip"
  tooltipProps={
    Object {
      "styles": Object {
        "content": Object {
          "display": "inline-block",
        },
        "root": Object {
          "lineHeight": "1.6rem",
        },
        "subText": Object {
          "color": "#ffffff",
          "fontSize": "1.2rem",
        },
      },
    }
  }
>
  text
</TooltipHost>
`;

exports[`<Tooltip /> with directionalHint matches its snapshot 1`] = `
<TooltipHost
  calloutProps={
    Object {
      "beakWidth": 8,
      "styles": Object {
        "beak": Object {
          "backgroundColor": "#343A41",
        },
        "beakCurtain": Object {
          "backgroundColor": "#343A41",
          "borderRadius": "2px",
        },
        "calloutMain": Object {
          "backgroundColor": "#343A41",
        },
        "container": Object {
          "backgroundColor": "inherit",
        },
        "root": Object {
          "backgroundColor": "inherit",
          "border": "none",
          "boxShadow": "none",
          "minHeight": "3.2rem",
          "padding": "0.7rem 1.2rem 0.8rem",
          "zIndex": 1000,
        },
      },
    }
  }
  content="text content"
  delay={1}
  directionalHint={4}
  hostClassName="y-tooltip"
  tooltipProps={
    Object {
      "styles": Object {
        "content": Object {
          "display": "inline-block",
        },
        "root": Object {
          "lineHeight": "1.6rem",
        },
        "subText": Object {
          "color": "#ffffff",
          "fontSize": "1.2rem",
        },
      },
    }
  }
/>
`;
